From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Omer Katz <omerkatz@chromium.org>
Date: Fri, 23 Oct 2026 10:12:28 +0200
Subject: [M156] [cppgc] Update the Steele-style barrier to always retrace

Original change's description:
> [cppgc] Update the Steele-style barrier to always retrace
>
> The barrier retraced objects only if they were already marked. This
> doesn't work well with concurrent marking because the object could be
> marked and the main thread hasn't observed it yet. Instead change the
> barrier to mark regardless of whether the object was already marked.
>
> Bug: 575143870
> Change-Id: I5f027c5c5c28cc23fc25d562d266081afe216dc1
> Reviewed-on: https://chromium-review.googlesource.com/c/v8/v8/+/8614239
> Reviewed-by: Michael Lippautz <mlippautz@chromium.org>
> Commit-Queue: Omer Katz <omerkatz@chromium.org>
> Cr-Commit-Position: refs/heads/main@{#111098}

(cherry picked from commit e19035b3162d3d998884c7363e75a7acc89133a5)

Bug: 577337481,575143870
Change-Id: I5f027c5c5c28cc23fc25d562d266081afe216dc1
Reviewed-on: https://chromium-review.googlesource.com/c/v8/v8/+/8667328
Bot-Commit: rubber-stamper@appspot.gserviceaccount.com <rubber-stamper@appspot.gserviceaccount.com>
Auto-Submit: chrome-cherry-picker@chops-service-accounts.iam.gserviceaccount.com <chrome-cherry-picker@chops-service-accounts.iam.gserviceaccount.com>
Commit-Queue: rubber-stamper@appspot.gserviceaccount.com <rubber-stamper@appspot.gserviceaccount.com>
Cr-Commit-Position: refs/branch-heads/15.6@{#51}
Cr-Branched-From: b976f35da8356fcf4e5e176d1d810ba6a6f05718-refs/heads/15.6.75@{#1}
Cr-Branched-From: a424b0597aab8e27f12ea4a700bae3961fa1a076-refs/heads/main@{#110135}

diff --git a/include/cppgc/heap-consistency.h b/include/cppgc/heap-consistency.h
index 23b5d909945b96db3bfba28fb4c66a25661914b8..31858f0a631bc85de50f8fdd8e881b9023d23b4c 100644
--- a/include/cppgc/heap-consistency.h
+++ b/include/cppgc/heap-consistency.h
@@ -146,8 +146,8 @@ class HeapConsistency final {
   }
 
   /**
-   * Steele-style write barrier that re-processes an object if it has already
-   * been processed.
+   * Steele-style write barrier that processes an object regardless of whether
+   * or not it has already been processed.
    *
    * \param params The parameters retrieved from `GetWriteBarrierType()`.
    * \param object The pointer to the object which must point to an object that
diff --git a/src/heap/cppgc-internal/marker.h b/src/heap/cppgc-internal/marker.h
index 84fec071262820633fbb8cd0f219b3fa1e2d5ec0..ad96e22f188b171721776fb38deba5c036ba0d02 100644
--- a/src/heap/cppgc-internal/marker.h
+++ b/src/heap/cppgc-internal/marker.h
@@ -10,6 +10,7 @@
 #include "include/cppgc/heap.h"
 #include "include/cppgc/platform.h"
 #include "include/cppgc/visitor.h"
+#include "src/base/logging.h"
 #include "src/base/macros.h"
 #include "src/base/platform/time.h"
 #include "src/heap/base/incremental-marking-schedule.h"
@@ -238,20 +239,15 @@ class V8_EXPORT_PRIVATE Marker final : public MarkerBase {
 
 template <MarkerBase::WriteBarrierType type>
 void MarkerBase::WriteBarrierForObject(HeapObjectHeader& header) {
-  // The barrier optimizes for the bailout cases:
-  // - kDijkstra: Marked objects.
-  // - kSteele: Unmarked objects.
-  switch (type) {
-    case MarkerBase::WriteBarrierType::kDijkstra:
-      if (!header.TryMarkAtomic()) {
-        return;
-      }
-      break;
-    case MarkerBase::WriteBarrierType::kSteele:
-      if (!header.IsMarked<AccessMode::kAtomic>()) {
-        return;
-      }
-      break;
+  // The Dijkstra barrier bails out on marked objects.
+  // The Steele barrier marks the object regardless of its current state.
+  // It's important to make sure the object is marked here since unmarking
+  // below for in construction objects expects a marked object.
+  const bool was_already_marked = !header.TryMarkAtomic();
+  if (type == MarkerBase::WriteBarrierType::kDijkstra) {
+    if (was_already_marked) {
+      return;
+    }
   }
 
   // The barrier fired. Filter out in-construction objects here. This possibly
@@ -267,13 +263,18 @@ void MarkerBase::WriteBarrierForObject(HeapObjectHeader& header) {
     return;
   }
 
-  switch (type) {
-    case MarkerBase::WriteBarrierType::kDijkstra:
-      mutator_marking_state_.write_barrier_worklist().Push(&header);
-      break;
-    case MarkerBase::WriteBarrierType::kSteele:
-      mutator_marking_state_.retrace_marked_objects_worklist().Push(&header);
-      break;
+  DCHECK_IMPLIES(was_already_marked,
+                 type == MarkerBase::WriteBarrierType::kSteele);
+  if ((type == MarkerBase::WriteBarrierType::kDijkstra) ||
+      !was_already_marked) {
+    // Push previously unmarked objects to the write barrier worklist, even for
+    // the Steele barrier, since popping from that worklist would also update
+    // the marked bytes counter.
+    mutator_marking_state_.write_barrier_worklist().Push(&header);
+  } else {
+    DCHECK((type == MarkerBase::WriteBarrierType::kSteele) &&
+           was_already_marked);
+    mutator_marking_state_.retrace_marked_objects_worklist().Push(&header);
   }
 }
 
diff --git a/test/unittests/heap/cppgc/write-barrier-unittest.cc b/test/unittests/heap/cppgc/write-barrier-unittest.cc
index 9df26a108ca6bce94f2759693a177500d1f7e0d6..bfe99ffa50e1e267b32e95e33b01e28cbc483b97 100644
--- a/test/unittests/heap/cppgc/write-barrier-unittest.cc
+++ b/test/unittests/heap/cppgc/write-barrier-unittest.cc
@@ -476,11 +476,11 @@ TEST_F(WriteBarrierTest, SteeleWriteBarrierTriggersWhenMarkingIsOn) {
   }
 }
 
-TEST_F(WriteBarrierTest, SteeleWriteBarrierBailoutIfNotMarked) {
+TEST_F(WriteBarrierTest, SteeleWriteBarrierTriggersEvenIfNotMarked) {
   auto* object1 = MakeGarbageCollected<GCed>(GetAllocationHandle());
   auto* object2 = MakeGarbageCollected<GCed>(GetAllocationHandle(), object1);
   {
-    ExpectNoWriteBarrierFires scope(marker(), {object1});
+    ExpectWriteBarrierFires scope(marker(), {object1});
     WriteBarrierParams params;
     EXPECT_EQ(WriteBarrierType::kMarking,
               HeapConsistency::GetWriteBarrierType(
